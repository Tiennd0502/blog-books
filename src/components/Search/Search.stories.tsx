import React from 'react';
import Storybook from '@storybook/react';

import Search from './index';

export default {
  title: 'Components/Search',
  component: Search,
  argTypes: {
    onSubmit: { action: 'onClick' },
  },
} as Storybook.ComponentMeta<typeof Search>;

const Template: Storybook.ComponentStory<typeof Search> = (args) => (
  <Search {...args} />
);

export const Default = Template.bind({});
Default.args = {
  placeholder: 'Search by title, author, or keyword',
};
