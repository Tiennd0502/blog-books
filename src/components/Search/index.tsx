import React, { useState } from 'react';
import { Input, InputGroup, InputRightElement } from '@chakra-ui/react';

import { SearchIcon } from '@chakra-ui/icons';

type Props = {
  placeholder: string;
  onSubmit: (value: string) => void;
};

const Search: React.FC<Props> = ({ placeholder, onSubmit }) => {
  const [keywords, setKeywords] = useState('');

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setKeywords(event.target.value);
  };

  const handleKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.keyCode === 13) {
      onSubmit(keywords);
    }
  };

  return (
    <InputGroup role="form">
      <Input
        pr="50px"
        pl="20px"
        minH="49px"
        color="yellow.900"
        borderRadius="30px"
        fontSize="24px"
        backgroundColor="blue.100"
        border="1px solid blue.100"
        type="text"
        value={keywords}
        onChange={handleInputChange}
        onKeyDown={handleKeyDown}
        placeholder={placeholder}
      />
      <InputRightElement w="50px" h="100%">
        <SearchIcon
          role="button"
          w="22px"
          h="22px"
          _hover={{ color: 'orange.500' }}
          onClick={() => onSubmit(keywords)}
        />
      </InputRightElement>
    </InputGroup>
  );
};

export default Search;
