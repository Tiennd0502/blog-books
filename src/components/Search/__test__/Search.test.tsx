import { render, fireEvent, cleanup } from '@testing-library/react';

import Search from '../index';

describe('Search component', () => {
  const PLACEHOLDER = 'Enter keywords...';
  const KEYWORDS = 'Lorem';
  const onSubmit = jest.fn();

  afterEach(cleanup);
  test('Should render Search component', () => {
    const { getByRole } = render(
      <Search placeholder={PLACEHOLDER} onSubmit={onSubmit} />
    );

    expect(getByRole('form')).toBeInTheDocument();
  });

  test('should simulate input search change', () => {
    const { getByPlaceholderText } = render(
      <Search placeholder={PLACEHOLDER} onSubmit={onSubmit} />
    );
    const input = getByPlaceholderText(PLACEHOLDER) as HTMLInputElement;
    fireEvent.change(input, { target: { value: KEYWORDS } });

    expect(input).toBeInTheDocument();
    expect(input.value).toBe(KEYWORDS);
  });

  test('should simulate input search on Enter', () => {
    const { getByPlaceholderText } = render(
      <Search placeholder={PLACEHOLDER} onSubmit={onSubmit} />
    );
    const input = getByPlaceholderText(PLACEHOLDER) as HTMLInputElement;
    fireEvent.change(input, { target: { value: KEYWORDS } });
    fireEvent.keyDown(input, { keyCode: 13 });
    onSubmit(KEYWORDS);

    expect(onSubmit).toHaveBeenCalled();
    expect(onSubmit).toHaveBeenCalledWith(KEYWORDS);
  });

  test('should simulate click button submit', () => {
    const { getByRole } = render(
      <Search placeholder={PLACEHOLDER} onSubmit={onSubmit} />
    );
    const btnSubmit = getByRole('button');
    fireEvent.click(btnSubmit);

    expect(onSubmit).toHaveBeenCalled();
  });

  test('Matches snapshot', () => {
    const { container } = render(
      <Search placeholder={KEYWORDS} onSubmit={onSubmit} />
    );

    expect(container).toMatchSnapshot();
  });
});
