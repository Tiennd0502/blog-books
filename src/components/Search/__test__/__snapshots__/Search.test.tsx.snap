// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component Matches snapshot 1`] = `
<div>
  <div
    class="chakra-input__group css-4302v8"
    role="form"
  >
    <input
      class="chakra-input css-n3on0e"
      placeholder="Lorem"
      type="text"
      value=""
    />
    <div
      class="chakra-input__right-element css-qoxuz7"
    >
      <svg
        class="chakra-icon css-18rs6r1"
        focusable="false"
        role="button"
        viewBox="0 0 24 24"
      >
        <path
          d="M23.384,21.619,16.855,15.09a9.284,9.284,0,1,0-1.768,1.768l6.529,6.529a1.266,1.266,0,0,0,1.768,0A1.251,1.251,0,0,0,23.384,21.619ZM2.75,9.5a6.75,6.75,0,1,1,6.75,6.75A6.758,6.758,0,0,1,2.75,9.5Z"
          fill="currentColor"
        />
      </svg>
    </div>
  </div>
</div>
`;
